src/crepe/system/InputSystem.cpp

```diff
@@ -0,0 +1,183 @@
+						.mouse_x = world_mouse_x,
+						.mouse_y = world_mouse_y,
+						.button = event.mouse_button,
+					});
+
+					handle_click(event.mouse_button, world_mouse_x, world_mouse_y);
+				}
+			} break;
+			case SDLContext::EventType::MOUSEMOVE:
+				if (!mouse_in_viewport) {
+					break;
+				}
+				event_mgr.queue_event<MouseMoveEvent>(MouseMoveEvent{
+					.mouse_x = world_mouse_x,
+					.mouse_y = world_mouse_y,
+					.rel_x = event.rel_mouse_move.first,
+					.rel_y = event.rel_mouse_move.second,
+				});
+				handle_move(event, world_mouse_x, world_mouse_y);
+				break;
+			case SDLContext::EventType::MOUSEWHEEL:
+				event_mgr.queue_event<MouseScrollEvent>(MouseScrollEvent{
+					.scroll_x = event.wheel_delta,
+					.scroll_y = 0,
+					.direction = event.wheel_delta,
+				});
+				break;
+			case SDLContext::EventType::SHUTDOWN:
+				event_mgr.queue_event<ShutDownEvent>(ShutDownEvent{});
+				break;
+			default:
+				break;
+		}
+	}
+}
+void InputSystem::handle_move(const SDLContext::EventData & event_data,
+							  const int & world_mouse_x, const int & world_mouse_y) {
+	ComponentManager & mgr = this->component_manager;
+
+	RefVector<Button> buttons = mgr.get_components_by_type<Button>();
+
+	for (Button & button : buttons) {
+		RefVector<Transform> transform_vec
+			= mgr.get_components_by_id<Transform>(button.game_object_id);
+		Transform & transform(transform_vec.front().get());
+
+		bool was_hovering = button.hover;
+		if (button.active
+			&& is_mouse_inside_button(world_mouse_x, world_mouse_y, button, transform)) {
+			button.hover = true;
+			if (!was_hovering && button.on_enter) {
+				button.on_enter();
+			}
+		} else {
+			button.hover = false;
+			// Trigger the on_exit callback if the hover state just changed to false
+			if (was_hovering && button.on_exit) {
+				button.on_exit();
+			}
+		}
+	}
+}
+
+void InputSystem::handle_click(const MouseButton & mouse_button, const int & world_mouse_x,
+							   const int & world_mouse_y) {
+	ComponentManager & mgr = this->component_manager;
+
+	RefVector<Button> buttons = mgr.get_components_by_type<Button>();
+
+	for (Button & button : buttons) {
+		RefVector<Transform> transform_vec
+			= mgr.get_components_by_id<Transform>(button.game_object_id);
+		Transform& transform(transform_vec.front().get());
+
+		if (button.active
+			&& is_mouse_inside_button(world_mouse_x, world_mouse_y, button, transform)) {
+			handle_button_press(button);
+		}
+	}
+}
+
+bool InputSystem::is_mouse_inside_button(const int & mouse_x, const int & mouse_y,
+										 const Button & button, const Transform & transform) {
+	int half_width = button.width / 2;
+	int half_height = button.height / 2;
+
+	return mouse_x >= transform.position.x - half_width
+		   && mouse_x <= transform.position.x + half_width
+		   && mouse_y >= transform.position.y - half_height
+		   && mouse_y <= transform.position.y + half_height;
+}
+
+void InputSystem::handle_button_press(Button & button) {
+	if (button.is_toggle) {
```

Comment:
Dit gedrag is nergens uitgelegd en naar mijn mening alleen verwarrend. Kan de `is_toggle` gewoon helemaal weg?

Reply:
Ik bemoei mij hier weer even. Qua mij betreft is de toggle optie wel erg handig, liever niet weggooien dus :).

Reply:
ik wilden het ook graag behouden ik kan proberen de functie wat duidelijker te maken, maar het doel van de engine is om het voor de gebruiker makkelijk te maken door functionaliteiten voor hun te maken. Ik denk dat de toggle zo'n geval is

Reply:
Ik ben soms onduidelijk met de feedback maar ik bedoelde met mijn originele comment dat het gedrag zoals de gebruiker deze ziet wanneer `is_toggle` aan staat nergens gedocumenteerd staat. De functie zelf is prima te volgen dus deze inline comments mogen weg.

Momenteel zorgt deze optie er dus alleen voor dat de helft van de `on_click` events genegeerd worden. Dit lijkt me nutteloos want dan kun je als programmeur alleen nog reageren op dat een optie aan moet gaan, maar je weet niet wanneer deze weer uit gaat. Als deze optie nog visueel iets zou aanpassen aan de knop waardoor de toestand zichtbaar wordt zou ik het begrijpen maar zelfs dan zou ik liever een losse ToggleButton klasse hebben met een `on_enable` en `on_disable` of `on_click(bool new_state)` inplaats van hoe het nu geïmplementeerd is.

Reply:
het is inderdaad onduidelijk ik kan deze voor nu weghalen en later een aparte toggle button kunnen toevoegen als derived class van button of ik kan de private variabelen op public zetten want dat kan de user wel zien of hij pressed is. een beetje zoals het bij css/javascript werkt